src/mutex.rs

```diff
@@ -259,158 +260,192 @@ impl<T: Default + ?Sized> Default for Mutex<T> {
+    }
 }
 
 unsafe impl<T: Send + ?Sized> Send for Lock<'_, T> {}
 unsafe impl<T: Sync + ?Sized> Sync for Lock<'_, T> {}
 
-impl<'a, T: ?Sized> Unpin for Lock<'a, T> {}
-
-impl<T: ?Sized> fmt::Debug for Lock<'_, T> {
+impl<T: ?Sized> fmt::Debug for LockInner<'_, T> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.write_str("Lock { .. }")
     }
 }
 
-impl<'a, T: ?Sized> Future for Lock<'a, T> {
+impl<'a, T: ?Sized> EventListenerFuture for LockInner<'a, T> {
     type Output = MutexGuard<'a, T>;
 
     #[inline]
-    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        let this = self.get_mut();
-
-        loop {
-            match this.acquire_slow.as_mut() {
+    fn poll_with_strategy<'x, S: event_listener_strategy::Strategy<'x>>(
+        self: Pin<&'x mut Self>,
+        strategy: &mut S,
+        context: &mut S::Context,
+    ) -> Poll<Self::Output> {
+        let mut this = self.project();
+
+        // This may seem weird, but the borrow checker complains otherwise.
+        if this.acquire_slow.is_none() {
+            match this.mutex.try_lock() {
+                Some(guard) => return Poll::Ready(guard),
                 None => {
-                    // Try the fast path before trying to register slowly.
-                    match this.mutex.try_lock() {
-                        Some(guard) => return Poll::Ready(guard),
-                        None => {
-                            this.acquire_slow = Some(AcquireSlow::new(this.mutex));
-                        }
-                    }
-                }
-
-                Some(acquire_slow) => {
-                    // Continue registering slowly.
-                    let value = ready!(Pin::new(acquire_slow).poll(cx));
-                    return Poll::Ready(MutexGuard(value));
+                    this.acquire_slow.set(Some(AcquireSlow::new(this.mutex)));
                 }
             }
         }
+
+        ready!(this
+            .acquire_slow
+            .as_pin_mut()
+            .unwrap()
+            .poll_with_strategy(strategy, context));
+        Poll::Ready(MutexGuard(this.mutex))
     }
 }
 
-/// The future returned by [`Mutex::lock_arc`].
-pub struct LockArc<T: ?Sized>(LockArcInnards<T>);
-
-enum LockArcInnards<T: ?Sized> {
-    /// We have not tried to poll the fast path yet.
-    Unpolled(Arc<Mutex<T>>),
+easy_wrapper! {
+    /// The future returned by [`Mutex::lock_arc`].
+    pub struct LockArc<T: ?Sized>(LockArcInnards<T> => MutexGuardArc<T>);
+    #[cfg(all(feature = "std", not(target_family = "wasm")))]
+    pub(crate) wait();
+}
 
-    /// We are acquiring the mutex through the slow path.
-    AcquireSlow(AcquireSlow<Arc<Mutex<T>>, T>),
+pin_project_lite::pin_project! {
+    #[project = LockArcInnardsProj]
+    enum LockArcInnards<T: ?Sized> {
+        /// We have not tried to poll the fast path yet.
+        Unpolled { mutex: Option<Arc<Mutex<T>>> },
```

Comment:
FYI, I think by using project_replace, we can eliminate the need for Option here and use the same way as before.